ai/core/src/unitycatalog/ai/core/utils/callable_utils.py

```diff
@@ -769,3 +772,209 @@ def check_docstring_signature_consistency(
+    """
+    type_str = type_str.strip()
+
+    if "<" not in type_str:
+        mapped = SQL_TYPE_TO_PYTHON_TYPE_MAPPING.get(type_str.upper())
+        return mapped.__name__ if mapped else type_str.lower()
+
+    outer, inner = type_str.split("<", 1)
+    outer = outer.strip().upper()
+    inner = inner.rsplit(">", 1)[0].strip()
+
+    parts = _split_generic_types(inner)
+
+    # recurse
+    parsed_parts = [_parse_sql_data_type(part) for part in parts]
+
+    # NB: For Python tuples, we map both tuple and list to the SQL ARRAY type.
+    # Since we can't disambiguate between them in the SQL type string, we assume that
+    # the collection container type is list for simplicity's sake, which is the first
+    # mapping entry in SQL_TYPE_TO_PYTHON_TYPE_MAPPING for ARRAY types.
+    if outer == "ARRAY":
+        return f"list[{parsed_parts[0]}]"
+    elif outer == "MAP":
+        return f"dict[{parsed_parts[0]}, {parsed_parts[1]}]"
+    elif outer == "STRUCT":
+        return "dict"
+    else:
+        return f"{outer.lower()}[{', '.join(parsed_parts)}]"
+
+
+def _reconstruct_docstring(function_info: "FunctionInfo", max_width: int = 100) -> str:
+    """
+    Reconstruct a Google-style docstring from a FunctionInfo object.
+
+    This function rebuilds:
+      - The function description from function_info.comment.
+      - An "Args:" section from each parameter's comment.
+      - A "Returns:" section using the return type (from full_data_type).
+
+    Each section is word-wrapped to a maximum width of `max_width` characters,
+    wrapping only at full words.
+
+    Returns:
+        A string representing the reconstructed docstring (with triple quotes and proper indentation).
+    """
+    doc_lines = []
+
+    # Overall function description.
+    if hasattr(function_info, "comment") and function_info.comment:
+        description = function_info.comment.strip()
+        wrapped_description = fill(description, width=max_width)
+        doc_lines.append(wrapped_description)
+
+    # Build the Args: section.
+    if function_info.input_params and function_info.input_params.parameters:
+        doc_lines.append("")  # Blank line.
+        doc_lines.append("Args:")
+        for param in function_info.input_params.parameters:
+            param_comment = (getattr(param, "comment", None) or "").strip()
+            arg_comment = f": {param_comment}" if param_comment else ""
+            arg_line = param.name + arg_comment
+            wrapped_arg = fill(
+                arg_line, width=max_width, initial_indent=" " * 4, subsequent_indent=" " * 6
+            )
+            doc_lines.append(wrapped_arg)
+
+    # Build the Returns: section.
+    return_type_str = _parse_sql_data_type(function_info.full_data_type)
+    doc_lines.append("")
+    doc_lines.append("Returns:")
+    wrapped_return = fill(
+        return_type_str, width=max_width, initial_indent="    ", subsequent_indent="    "
```

Comment:
Since "    " and " "*4 are used in many places in this codebase, shall we make it a const?

Reply:
Excellent call out :) Updated!

Reply:
ooof yeah. Thanks! Missed that one and the other docstring formatting one. Updated them and will push soon